### PR TITLE
Turrets

### PR DESCRIPTION
**Pull request by Zzarek**
Syncing for UITurret_Window

processors, patch, and packets for all of the following functionality.

- Ammo Feeding
- Turret Group
- Burst Mode
- Priority Configuration

Also includes 
- fix for statistics window
- stub for SuperNova. This is Probably worth throwing out some of, but wanted to open for discussion if any of it is useful first. At the very least it could be tailored into a starting spot.

(Ignore the branch name, I started on battlebase, but changed gear)

### CODE REVIEW

File: NebulaNetwork/PacketProcessors/Factory/Turret/TurretStorageUpdateProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        var pool = GameMain.galaxy.PlanetById(packet.PlanetId)?.factory?.defenseSystem.turrets;
+        if (pool == null || packet.TurretIndex == -1 || packet.TurretIndex > pool.buffer.Length ||
+            pool.buffer[packet.TurretIndex].id == -1)
+        {
+            return;
+        }
+
+        if (packet.ItemCount == 0 || packet.ItemId == 0)
+        {
+            pool.buffer[packet.TurretIndex].ClearItem();
```
Comment (Therzok) [inline NebulaNetwork/PacketProcessors/Factory/Turret/TurretStorageUpdateProcessor.cs:27]: Would help a bit at runtime if this was a local variable. Avoids a memory load/store

**Review by Therzok — APPROVED** [review]

File: NebulaNetwork/PacketProcessors/Factory/Turret/TurretStorageUpdateProcessor.cs (hunk 1 above)
Comment (Zzarek) [reply NebulaNetwork/PacketProcessors/Factory/Turret/TurretStorageUpdateProcessor.cs:27]: I followed the existing pattern of the app, all processors are done this way.  Any performance tuning is always a good thing, but code consistency, especially in an app with multiple maintainers, reigns king in my opinion. Well at least so long as the performance isn't an issue.

**Review by starfi5h — APPROVED** [review]

File: NebulaPatcher/Patches/Dynamic/UITurretWindow_Patch.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        //Notify about manual bullet inserting / withdrawing change
+        if (!Multiplayer.IsActive)
+        {
+            return;
+        }
+        var turret = __instance.defenseSystem.turrets.buffer[__instance.turretId];
+
+        Multiplayer.Session.Network.SendPacketToLocalStar(new TurretStorageUpdatePacket(turret,
+            GameMain.localPlanet?.id ?? -1));
+    }
```
Comment (PhantomGamers) [inline NebulaPatcher/Patches/Dynamic/UITurretWindow_Patch.cs:27]:
```suggestion
    [HarmonyPostfix]
    [HarmonyPatch(nameof(UITurretWindow.OnHandFillAmmoButtonClick))]
    public static void OnHandFillAmmoButtonClick_Postfix(UITurretWindow __instance)
    {
        //Notify about manual bullet inserting / withdrawing change
        if (!Multiplayer.IsActive || __instance is null)
        {
            return;
        }

        var defenseSystem = __instance.defenseSystem;

        // UITurretWindow closed
        if (defenseSystem is null)
        {
            return;
        }

        var turret = defenseSystem.turrets.buffer[__instance.turretId];

        Multiplayer.Session.Network.SendPacketToLocalStar(new TurretStorageUpdatePacket(turret,
            GameMain.localPlanet?.id ?? -1));
    }
```

File: NebulaNetwork/PacketProcessors/Factory/Turret/TurretGroupUpdateProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+{
+    protected override void ProcessPacket(TurretGroupUpdatePacket packet, NebulaConnection conn)
+    {
+        var pool = GameMain.galaxy.PlanetById(packet.PlanetId)?.factory?.defenseSystem.turrets;
+        if (pool != null && packet.TurretIndex != -1 && packet.TurretIndex < pool.buffer.Length &&
+            pool.buffer[packet.TurretIndex].id != -1)
+        {
+            pool.buffer[packet.TurretIndex].SetGroup(packet.Group);
+        }
+    }
```
Comment (PhantomGamers) [inline NebulaNetwork/PacketProcessors/Factory/Turret/TurretGroupUpdateProcessor.cs:23]:
```suggestion
    protected override void ProcessPacket(TurretGroupUpdatePacket packet, NebulaConnection conn)
    {
        var pool = GameMain.galaxy.PlanetById(packet.PlanetId)?.factory?.defenseSystem.turrets;
        if (pool == null || packet.TurretIndex == -1 || packet.TurretIndex >= pool.buffer.Length)
        {
            return;
        }
        var buffer = pool.buffer[packet.TurretIndex];
        if (buffer.id != -1)
        {
            pool.buffer[packet.TurretIndex].SetGroup(packet.Group);
        }
    }
```

**Pull request merged**